rnd/autogpt_server/autogpt_server/executor/manager.py

```diff
@@ -231,41 +229,62 @@ class Executor:
     loop: asyncio.AbstractEventLoop
 
     @classmethod
-    def on_executor_start(cls):
+    def on_node_executor_start(cls):
         cls.loop = asyncio.new_event_loop()
         cls.loop.run_until_complete(db.connect())
         cls.agent_server_client = get_agent_server_client()
 
     @classmethod
-    def on_start_execution(cls, q: ExecutionQueue, data: Execution) -> bool:
+    def on_node_execution(cls, q: ExecutionQueue[NodeExecution], data: NodeExecution):
         prefix = get_log_prefix(data.graph_exec_id, data.node_exec_id)
         try:
-            logger.warning(f"{prefix} Start execution")
+            logger.warning(f"{prefix} Start node execution")
             for execution in execute_node(cls.loop, cls.agent_server_client, data):
                 q.add(execution)
-            return True
+            logger.warning(f"{prefix} Finished node execution")
         except Exception as e:
             logger.exception(f"{prefix} Error: {e}")
-            return False
+
+    @classmethod
+    def on_graph_executor_start(cls):
+        cls.pool_size = Config().num_node_workers
+        cls.executor = ProcessPoolExecutor(
+            max_workers=cls.pool_size,
+            initializer=cls.on_node_executor_start,
+        )
+        logger.warning(f"Graph executor started with max-{cls.pool_size} node workers.")
+
+    @classmethod
+    def on_graph_execution(cls, graph_data: GraphExecution):
+        queue = ExecutionQueue[NodeExecution]()
+        for node_exec in graph_data.start_node_execs:
+            queue.add(node_exec)
+
+        futures: list[Future] = []
+        while not queue.empty():
+            futures.append(
+                cls.executor.submit(cls.on_node_execution, queue, queue.get())
+            )
+            while queue.empty() and len(futures) > 0:
+                future = futures.pop()
+                future.result()
 
 
 class ExecutionManager(AppService):
-    def __init__(self, pool_size: int):
-        self.pool_size = pool_size
-        self.queue = ExecutionQueue()
+    def __init__(self):
+        self.pool_size = Config().num_graph_workers
```

Comment:
@ntindle I want to get the config on a newly created process without initializing the secret too, is this okay?

Reply:
that is okay and a large reason why they are split